### PR TITLE
Glue: Enrich query results data frames in Explore with correlations to generate static links from correlations

### LINKED ISSUE #55370: Glue: Enrich query results data frames in Explore with correlations to generate static links from correlations
_Requirements_

When a new static query is added to correlations configuration,
And the user opens a panel that loads a configured data link
Then clicking on the data link button opens static query provided in the correlations config

_Suggested implementation_

(1) Add a new decorator to enrich query results in Explore in https://github.com/grafana/grafana/blob/df015a9301baf4b85eb321e6fc900c753891bcfc/public/app/features/explore/utils/decorators.ts#L161

(2) Use `useCorrelations` to create `correlationsConfig` (map refId to correlations config). The config should be passed to `runQueries()` action in Explore.

Pseudo-code for enriching

```typescript
// e.g.
type CorrelationsFieldConfig = {
  [ fieldName: string ] : DataLink[];
}
type CorrelationsConfig = {
 [ queryRefId: string ] : CorrelationsFieldConfig[];
}
// e.g.: config: CorrelationsConfig = { A: { fieldA: [], fieldB: [] }, B: ... }

export function decorateData(
  data: PanelData,
  queryResponse: PanelData, // contains data frames with refId referencing the query (no reference to data source)
  absoluteRange: AbsoluteTimeRange,
  refreshInterval: string | undefined,
  queries: DataQuery[] | undefined, // contains data source reference
  fullRangeLogsVolumeAvailable: boolean;
  correlationsConfig: CorrelationsConfig;
) {
  // ...
  map(decorateWithCorrelations)
}

function decorateWithCorrelations(queryResponse, correlationsConfig) {
const dataFrames = queryResponse.series; 
  dataFrames.forEach( (frame) => {
    const queryRefId = frame.refId;
    // enrich frame based on correlationsConfig
}
```

### PR DESCRIPTION
**Pull request by ifrost**
**What this PR does / why we need it**:

Creating generic data links based on the defined correlations. Correlations are loaded every time Explore page is opened.

It requires a feature toggle in `custom.ini`:

```ini
[feature_toggles]
correlations = true
```

The provisioning file for `devenv` defines a correlation in `gdev-loki` attached to `traceId` (which is a derived field). It can be tested with `make devenv sources=loki`:

```yaml
  - name: gdev-loki
    ...
    correlations:
      ...
      - targetUID: gdev-prometheus
        label: "Logs to metrics"
        description: "Related metrics stored in Prometheus"
        config:
          type: query
          target:
            expr: "{ job=\"test\" }"
          field: "traceID"
```

https://user-images.githubusercontent.com/745532/193818912-96001391-e6e5-44e4-b1c3-73dadc3df748.mov

**Which issue(s) this PR fixes**:

Fixes #55370

**Special notes for your reviewer**:

<details><summary>How to test it</summary>

* Define a correlation in provisioning file
* Run a log query that returns a field defined in the correlation
* The button to run attached query should be shown with the label of the correlation
* Clicking on the button should run attached query

</details>

### CODE REVIEW

File: public/app/features/explore/Wrapper.tsx [hunk 1]
```diff
@@ -30,10 +31,12 @@ const styles = {
 
 function Wrapper(props: GrafanaRouteComponentProps<{}, ExploreQueryParams>) {
   useExplorePageTitle();
+  const { correlations } = useSelector((state) => state.explore);
   const dispatch = useDispatch();
   const queryParams = props.queryParams;
-  const { keybindings, chrome } = useGrafana();
+  const { keybindings, chrome, config } = useGrafana();
   const navModel = useNavModel('explore');
+  const { get } = useCorrelations();
```
Comment (gelicia) [inline public/app/features/explore/Wrapper.tsx:39]: can we alias this to something more specific?

File: public/app/features/explore/Wrapper.tsx [hunk 2]
```diff
@@ -72,6 +90,10 @@ function Wrapper(props: GrafanaRouteComponentProps<{}, ExploreQueryParams>) {
 
   const hasSplit = Boolean(queryParams.left) && Boolean(queryParams.right);
 
+  if (!correlations) {
```
Comment (gelicia) [inline public/app/features/explore/Wrapper.tsx:93]: maybe premature optimization, but I'm wondering if we would really want to keep the whole rapper from showing until its loaded all the correlations? would we want this to be per datasource?

**Review by Elfo404 — CHANGES_REQUESTED** [review]
not sure why it doesn't show conflicts here, but i think a merge with main is due as (AFAICT) this branch is based on the wrapper revision that included code for split resize

**Comment by Elfo404** [comment]
My bad. i didn't update my local branch :pray:

File: public/app/features/explore/Wrapper.tsx (hunk 2 above)
Comment (ifrost) [reply public/app/features/explore/Wrapper.tsx:93]: Had similar thoughts. In theory we could load correlations only after getting results so we don't even need to delay running the queries. Just went with the simplest option for now to see how it evolves, but can be optimised now as well.

File: public/app/features/explore/Wrapper.tsx (hunk 1 above)
Comment (ifrost) [reply public/app/features/explore/Wrapper.tsx:39]: Hmm... maybe we should rename it if we always want to alias it?

File: public/app/features/explore/Wrapper.tsx (hunk 2 above)
Comment (ifrost) [reply public/app/features/explore/Wrapper.tsx]: @Elfo404 / @gelicia would something like this https://github.com/grafana/grafana/pull/56295/commits/ec0bd25bde978debe280635f17135e17ad289a25 make sense 🤔

Comment (Elfo404) [reply public/app/features/explore/Wrapper.tsx]: yasss ❤️ 

not sure about the naming of `ensureCorrelations` but i also wouldn't know how to call it :D

Comment (ifrost) [reply public/app/features/explore/Wrapper.tsx]: Thanks for suggesting it 👍  it's basically your idea sprinkled with some stupid function names ❤️

File: public/app/features/explore/state/query.ts [hunk 3]
```diff
@@ -904,6 +912,28 @@ export const queryReducer = (state: ExploreItemState, action: AnyAction): Explor
   return state;
 };
 
+/**
+ * Creates an observable that emits correlations once they are loaded
+ */
+const ensureCorrelations = () => {
```
Comment (Elfo404) [inline public/app/features/explore/state/query.ts:918]: nit: what about
```suggestion
const getCorrelations$ = () => {
```
?

seems to be more inline with rxjs naming conventions

File: public/app/features/explore/utils/decorators.ts [hunk 4]
```diff
@@ -77,6 +79,22 @@ export const decorateWithFrameTypeMetadata = (data: PanelData): ExplorePanelData
...
+  queries,
+  correlations,
+}: {
+  queries: DataQuery[] | undefined;
+  correlations: CorrelationData[] | undefined;
+}) => {
+  return (data: PanelData): PanelData => {
+    if (!isEmpty(queries) && !isEmpty(correlations)) {
+      const queryRefIdToDataSourceUid = mapValues(groupBy(queries, 'refId'), '0.datasource.uid');
+      attachCorrelationsToDataFrames(data.series, correlations!, queryRefIdToDataSourceUid);
```
Comment (Elfo404) [inline public/app/features/explore/utils/decorators.ts:92]: i think in this case not using `isEmpty` may be better as we then don't have to do a non-null assertion on `correlations`
```suggestion
    if (queries?.length && correlations?.length) {
      const queryRefIdToDataSourceUid = mapValues(groupBy(queries, 'refId'), '0.datasource.uid');
      attachCorrelationsToDataFrames(data.series, correlations, queryRefIdToDataSourceUid);
```

**Review by Elfo404 — APPROVED** [review]
Works great!

https://user-images.githubusercontent.com/1170767/195076077-5e1e1c74-e374-4df3-8851-38c618bcea0c.mov

left a couple of comments that are more nits than anything else, feel free to ignore if you don't agree

File: public/app/features/explore/state/query.ts (hunk 3 above)
Comment (ifrost) [reply public/app/features/explore/state/query.ts]: Fixed in ddaac692a3de1fe3dcfbebb4a6c4dafadcb757bf

File: public/app/features/explore/utils/decorators.ts (hunk 4 above)
Comment (ifrost) [reply public/app/features/explore/utils/decorators.ts]: Fixed in d4ac82b4f240271dab7a3490827bdf59d079ee90

**Pull request merged**